File: ga/4omini_result_2024_7/result/1968D_mini/0_1968D.c
use std::io;

fn min(a: usize, b: usize) -> usize {
    if a < b {
        a
    } else {
        b
    }
}

fn maxl(a: i64, b: i64) -> i64 {
    if a > b {
        a
    } else {
        b
    }
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();

    for _ in 0..t {
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        let params: Vec<usize> = input.split_whitespace()
                                       .map(|s| s.parse().unwrap())
                                       .collect();
        let (n, k, a, b) = (params[0], params[1], params[2], params[3]);
        let m = min(n, k);
        
        let mut arr = vec![0; n + 1];
        let mut aa = vec![0; n + 1];

        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        for (i, value) in input.split_whitespace().enumerate() {
            arr[i + 1] = value.parse().unwrap();
        }

        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        for (i, value) in input.split_whitespace().enumerate() {
            aa[i + 1] = value.parse().unwrap();
        }

        let mut p1 = 0;
        let mut p2 = 0;
        let mut p1_ = 0;
        let mut p2_ = 0;

        let mut a_index = a;
        let mut b_index = b;

        for x in 1..=m {
            p1 = maxl(p1, (aa[a_index] * (k - x + 1) as i64) + p1_);
            p1_ += aa[a_index];
            a_index = arr[a_index];

            p2 = maxl(p2, (aa[b_index] * (k - x + 1) as i64) + p2_);
            p2_ += aa[b_index];
            b_index = arr[b_index];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}

File: ga/gpt3_5_result_2024_6/result/1968D_gpt/0_1968D.rs
use std::cmp::{max, min};

const N: usize = 200000;

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();

    for _ in 0..t {
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut parts = input.trim().split_whitespace();
        let n: usize = parts.next().unwrap().parse().unwrap();
        let k: usize = parts.next().unwrap().parse().unwrap();
        let a: usize = parts.next().unwrap().parse().unwrap();
        let b: usize = parts.next().unwrap().parse().unwrap();
        let m = min(n, k);
        let mut a_ = a;
        let mut b_ = b;
        let mut arr: Vec<usize> = vec![0; n + 1];
        let mut aa: Vec<usize> = vec![0; n + 1];

        let mut input = String::new();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut parts = input.trim().split_whitespace();
        for x in 1..=n {
            arr[x] = parts.next().unwrap().parse().unwrap();
        }

        let mut input = String::new();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut parts = input.trim().split_whitespace();
        for x in 1..=n {
            aa[x] = parts.next().unwrap().parse().unwrap();
        }

        let mut p1: i64 = 0;
        let mut p2: i64 = 0;
        let mut p1_: i64 = 0;
        let mut p2_: i64 = 0;

        for x in 1..=m {
            p1 = max(p1, 1i64 * aa[a] * (k - x + 1) + p1_);
            p1_ += aa[a];
            a = arr[a];
            p2 = max(p2, 1i64 * aa[b] * (k - x + 1) + p2_);
            p2_ += aa[b];
            b = arr[b];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}